### ISSUE TITLE
Task Assignment

### ISSUE THREAD

**Issue by zlavergne** [hotosm/tasking-manager#1575]
I have implemented task assignment internally for FB. 
Related issue: #1309 

Implementation:
* Optional setting to enable task assignment
<img width="1100" alt="Screen Shot 2019-05-14 at 6 41 37 PM" src="https://user-images.githubusercontent.com/9849118/57741018-efed4180-7677-11e9-957d-e14e4651294f.png">

* The map shows icons to show which tasks are assigned to the current user and others (green icon for current/grey for others)
<img width="1179" alt="Screen Shot 2019-05-14 at 6 43 02 PM" src="https://user-images.githubusercontent.com/9849118/57741050-21660d00-7678-11e9-9b1a-dde68d171684.png">

* PM/Admins can manually assign and un-assign users to a task
![May-14-2019 18-52-33](https://user-images.githubusercontent.com/9849118/57741303-7e15f780-7679-11e9-8b2f-170f2ea21922.gif)

* Users can see tasks assigned to them...
<img width="1272" alt="Screen Shot 2019-05-14 at 6 47 43 PM" src="https://user-images.githubusercontent.com/9849118/57741219-ff20bf00-7678-11e9-9a2d-6121c94c51ae.png">

* And tasks they've assigned to others
<img width="1246" alt="Screen Shot 2019-05-14 at 6 47 53 PM" src="https://user-images.githubusercontent.com/9849118/57741227-08119080-7679-11e9-9ff2-cddced52621c.png">

* If a task is invalidated, it is re-assigned back to the user who last mapped the task
![May-14-2019 18-57-54](https://user-images.githubusercontent.com/9849118/57741483-452a5280-767a-11e9-8edf-524ba90f58d8.gif)

**Comment by xamanu** [comment]
This is very nice! I would like to get this into the Tasking Manager. Perhaps not the usual volunteer workflow, but seems to be a relevant functionality for very organized mapping teams.

**Comment by russdeffner** [comment]
@xamanu @zlavergne - From my perspective, it would be good to be able to Enforce Random Selection for public projects as well. But only if it is coupled with the random selection reflecting priority areas and/or tasks. I could see wanting to use it for disaster activation in a similar manner where I don't want people spending time on the 'easy' (often the coast or edge of project where there's very little to map), but focusing on the priority areas where data needs are most critical.

**Comment by abalosc1** [comment]
The task assignment workflow could also be good for making sure tasks get assigned to a validator.

It would also be cool to be able to select multiple tasks to assign (like how you can draw a polygon over tasks to validate).

And to be able to see what tasks a selected user has been assigned (so you don't end up giving that user too many tasks). Here was our TM2 implementation that had the selected user's tasks highlighted in green and tasks assigned to other users highlighted in gray. The assignments on each task could be seen on hover.
![assigned to user](https://user-images.githubusercontent.com/34665975/57813852-a1a37580-773f-11e9-9cb9-5f4f0e22eb0f.jpg)

@zlavergne What do the green and gray person icons mean in your screenshots?

**Comment by xamanu** [comment]
@zlavergne, I think the enforce random selection feature is very valuable, especially if combined with priority areas, as @russdeffner suggested. However I would separate the enforce random selection from this issue. @zlavergne, do you think this is possible without too much overhead?

**Comment by zlavergne** [comment]
Per @xamanu's suggestion, I move the random task selection conversation to #1583

**Comment by zlavergne** [comment]
@abalosc1 I agree it would be great to be able to select multiple (my API is built to handle multiple tasks). Maybe a `shift+click` would be nice too to allow for some specific assignments. Or do you think a polygon would be better? 

> What do the green and gray person icons mean in your screenshots?

green => assigned to you (current user)
grey => assigned to someone else

**Comment by abalosc1** [comment]
> Maybe a `shift+click` would be nice too to allow for some specific assignments. Or do you think a polygon would be better?
> 
> >

Yes, I think clicking is more intuitive/easier than trying to draw a polygon around the tasks to select them

**Comment by zlavergne** [comment]
@russdeffner how do you think your use cases would use task assignment? If "task assignment" is too aggressive it can also be thought of as "task ownership".

**Comment by russdeffner** [comment]
@zlavergne Yes, I could see use cases for HOT to use the task assignment. Internal teams I would think anyone utilizing organized mapping teams could benefit from the ability to assign tasks. Only thing I am unsure about, would it allow for assigning for each stage? i.e. I want to assign to mapper X, then validator Y?

**Comment by abalosc1** [comment]
> Only thing I am unsure about, would it allow for assigning for each stage? i.e. I want to assign to mapper X, then validator Y?

We've gotten requests from our users to unassign mappers once they've marked the task as done. Then the task is open to be assigned to someone else like a validator.

**Comment by zlavergne** [comment]
> Only thing I am unsure about, would it allow for assigning for each stage? i.e. I want to assign to mapper X, then validator Y?

It's designed to do just that. 

There's 3 possible ways for a task to be assigned:
1. PM/Admin manually assigns tasks to a user
2. Task is automatically assigned to a user when they lock the task 
3. When a task is marked `invalid` it is re-assigned to the user who marked it as `mapped`

And there's 2 ways for it to be unassigned:
1. PM/Admin manually un-assigns tasks
2. User marks task as complete

> We've gotten requests from our users to unassign mappers once they've marked the task as done. 

In this version, tasks are automatically un-assigned when marked as done.
